design/vela-core/environment.md

```diff
@@ -0,0 +1,144 @@
+
+Here are some examples of what users want to define in an environment:
+
+- K8s Cluster. *dev* environment might have smaller clusters while *prod* might use large and complex clusters.
+- System components and definitions. *dev* environment might install different Operators & CRDs from *prod* environment, e.g. *dev* might choose local logging while *prod* might use ELK logging.
+- Network boundary. *dev* and *prod* environments would have different VPC setups, gateway endpoints, and firewall settings.
+- Secret and config. *dev* might use locally-run MySQL, while *prod* might use managed RDS. The DB connection credentials would be saved differently.
+- Policies. This is more important in *prod* environment that it would usually have more restrictions on application deployments, like security scanning, config checks, SLO checks, etc.
+
+All of above examples and more are bundled as a shared base for applications to run on.
+This is usualy done by an admin for the entire application team.
+By adding Environment API, Vela aligns with user workflow better,
+provides consistent view and pluggable options for end users,
+and empowering platform builders to add modular environment capabilities. In the following, we will propose the design of API and give an implementation plan.
+
+
+## Proposal
+
+### 1. Environment CRD
+
+We propose to add an Environment CRD to describe the shared resources for applications to base on.
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # A list of pluggable EnvConfig resources.
+  # The resources will be parsed and applied when Environment CR is appled.
+  configs:
+    - type: cluster
+      properties:
+        name: prod-cluster
+        selector:
+          tier: prod
+
+    - type: system-setup
+      properties:
+        components:
+          - type: helm-git
+            properties:
+              git: git-repo-url
+              path: path/to/chart
+              valuePath: value.yaml
+          - type: kustomize-git
+            properties:
+              kube:
+                git: git-repo-url
+                path: path/to/dir
+    
+    - type: secrets
+      properties:
+        redis:
+          url: redis-url
+          password: redis-password
+    
+    - type: policies
+      properties:
+        quota: ...
+        security: ...
+        slo: ...
+```
+
+We also need to add an optional field `environment` in Application:
+
+```yaml
+kind: Application
+spec:
+  # If set, app controller will add the AppRevision into the Environment
+  # CR spec after generating AppRevision before executing Workflow.
+  environment: prod
+```
+
+### 2. EnvConfigDefinition CRD
+
+An Environment can have multiple EnvConfigs. The schema is defined in EnvConfigDefinition CRD:
+
+```yaml
+apiVersion: core.oam.dev/v1beta1
+kind: EnvConfigDefinition
```

Comment:
Can we merge it with PolicyDefinition/WorkflowStepDefinition and call them Definition? We should reduce CRD to reduce complexity.

Reply:
What's your suggestion? How can we do that if they are for different things and have different spec fields?

Reply:
Let's make a unified Object called CUEPackage, and the policy,workflow and env all rely on the CUEPackage as abstraction. And in the implementation, they're all CUE based templates.

Reply:
First of all, I don't think we should only make policy, workflow, and env. If we want to make a unified object, all definitions should be there, including component and trait. We shouldn't treat policy, etc. as separate definitions.

Second, what about the specific fields that policy, workflow, or env have? They need to be in spec.